### ISSUE TITLE
[BUG] Weird behaviour of `DPODataCollatorWithPadding`

### ISSUE THREAD

**Issue by nrwugh** [huggingface/trl#907]
# Missing BOS token

In SFT stage of stackllama2 demo, bos token ids are added(since llama tokenizer adds BOS token by default).
https://github.com/huggingface/trl/blob/main/trl/trainer/utils.py#L574
```python
            tokenized_inputs = self.tokenizer(buffer, truncation=False)["input_ids"]
            all_token_ids = []
            for tokenized_input in tokenized_inputs:
                all_token_ids.extend(tokenized_input + [self.concat_token_id])
```

But in DPO stage, special tokens are not added.
https://github.com/huggingface/trl/blob/main/trl/trainer/utils.py#L321-L323
```python
            chosen_tokens = self.tokenizer(chosen, add_special_tokens=False)
            rejected_tokens = self.tokenizer(rejected, add_special_tokens=False)
            prompt_tokens = self.tokenizer(prompt, add_special_tokens=False)
```

Later EOS tokens are added, while BOS tokens are left...
https://github.com/huggingface/trl/blob/main/trl/trainer/utils.py#L347-L353
```python
            # add EOS token to end of prompt
            chosen_tokens["input_ids"].append(self.tokenizer.eos_token_id)
            chosen_tokens["attention_mask"].append(1)

            rejected_tokens["input_ids"].append(self.tokenizer.eos_token_id)
            rejected_tokens["attention_mask"].append(1)
```

I believe this is a bug, since it causes inconsistent behaviour in 2 stages.

# (Maybe) unnecessary new attention mask

Since `add_special_tokens=False`, there should be no eos token id in xxx_tokens.
But later the code try to mark eos id and make its attention mask 0.
Isn't that unnecessary?

https://github.com/huggingface/trl/blob/main/trl/trainer/utils.py#L321C13-L332
```python
            chosen_tokens = self.tokenizer(chosen, add_special_tokens=False)
            rejected_tokens = self.tokenizer(rejected, add_special_tokens=False)
            prompt_tokens = self.tokenizer(prompt, add_special_tokens=False)

            eos_token_id = self.tokenizer.eos_token_id
            # Get indices in list prompt_tokens["input_ids"] that equals the EOS token (often 0)
            eos_indices_prompt = [i for i, x in enumerate(prompt_tokens["input_ids"]) if x == eos_token_id]
            # attention mask these indices to eos_token_id
            new_attention_mask = [
                0 if i in eos_indices_prompt else p for i, p in enumerate(prompt_tokens["attention_mask"])
            ]
            prompt_tokens["attention_mask"] = new_attention_mask
```

**Comment by lvwerra** [comment]
cc @kashif

**Comment by nrailg** [comment]
# Should not concat token ids of 2 strings

Llama Tokenizer uses SPM. Space could be merged with word behind and form another different token.

```
print('A', tokenizer('Answer: ').input_ids)
print('B', tokenizer('A').input_ids)
print('C', tokenizer('Answer: A').input_ids)
```

```
A [1, 673, 29901, 29871]
B [1, 319]
C [1, 673, 29901, 319]
```

**Comment by kashif** [comment]
yes @nrailg  we encourage community fixes so please go ahead!
